### PR TITLE
Add deployment options for individual yaml files

### PR DESCRIPTION
**Pull request by loosebazooka**

### CODE REVIEW

File: src/main/java/com/google/cloud/tools/appengine/api/deploy/AppEngineDeployment.java [hunk 1]
```diff
@@ -31,4 +31,28 @@
    */
   void deploy(DeployConfiguration config) throws AppEngineException;
 
+  /**
+   * Deploy cron configuration to App Engine.
+   */
+  void deployCron(DeployCronConfiguration config) throws AppEngineException;
+
+  /**
+   * Deploy dos configuration to App Engine.
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/api/deploy/AppEngineDeployment.java]: dos? disk operating system? :-)

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkAppEngineDeployment.java [hunk 2]
```diff
@@ -93,4 +98,99 @@ public void deploy(DeployConfiguration config) throws AppEngineException {
...
+    } catch (ProcessRunnerException e) {
+      throw new AppEngineException(e);
+    }
+  }
+
+  @Override
+  public void deployQueue(DeployQueueConfiguration config) throws AppEngineException {
+    Preconditions.checkNotNull(config);
+    Preconditions.checkNotNull(config.getQueueYaml());
+    Preconditions.checkArgument(config.getQueueYaml().getName().equals("queue.yaml"),
```
Comment (chanseokoh) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkAppEngineDeployment.java]: Are we forcing users to use a fixed name for these files? Is this what the way it works?

Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkAppEngineDeployment.java]: yeah, to deply a cron.yaml and update a project's cron config

```
gcloud app deploy cron.yaml
```

The only name you can change is `app.yaml`

**Comment by loosebazooka** [comment]
@meltsufin yeah this feels a little redundant, but I believe the method you suggest would be less strict in how we enforce naming. The explicit definition could potentially lead to fewer user errors. Unless I misunderstand what you are suggesting.

Also when defining the configuration in plugins, I suspect we don't want to reuse configuration names. If we going to do cronYaml anyway (in gradle or maven for example) and map it to a generic yaml name when configuring, we're moving that work out of the library and into the client.

**Comment by meltsufin** [comment]
@loosebazooka I see what you're saying. In that case, maybe create a `DeployYamlConfiguration` and then subclass it? This way you can still just have one method in `CloudSdkAppEngineDeployment` called `deployYaml`?

**Comment by loosebazooka** [comment]
@meltsufin I want to make the names explicit for two reasons

1. so that in the maven/gradle config they are explicit
```
deploy {
  cronYaml = "path/to/cron.yaml"
}
```
and the handler in the client (build tool) just passes that on directly.

2. we're doing checks on the naming to see if they are deploying a "cron.yaml", for instance if they did
deploy {
  cronYaml = "path/to/app.yaml"
}
and we would explicitly fail here in the preconditions.

**Comment by loosebazooka** [comment]
@meltsufin moved some code around for better re-use, also rebased against master. ptal

**Comment by meltsufin** [comment]
@loosebazooka looks good - definitely a reduction in code duplication.

**Comment by loosebazooka** [comment]
I might take another look here, do it more like Mike suggested as a generic method

**Comment by loosebazooka** [comment]
@meltsufin I completely rewrote this after having a hard time convincing myself to submit it. PTAL.

File: src/main/java/com/google/cloud/tools/appengine/api/deploy/DeployProjectConfigurationConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package com.google.cloud.tools.appengine.api.deploy;
+
+import com.google.cloud.tools.appengine.api.Configuration;
+
+import java.io.File;
+
+/**
+ * Created by appu on 3/21/17.
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/api/deploy/DeployProjectConfigurationConfiguration.java]: You need this?

File: src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultDeployProjectConfigurationConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.appengine.api.deploy;
+
+import com.google.cloud.tools.appengine.api.DefaultConfiguration;
+
+import java.io.File;
+
+public class DefaultDeployProjectConfigurationConfiguration extends DefaultConfiguration implements
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultDeployProjectConfigurationConfiguration.java]: `ConfigurationConfiguration`.. really? :)

File: src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultDeployProjectConfigurationConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    DeployProjectConfigurationConfiguration {
+
+  private File projectConfigurationDirectory;
+
+  @Override
+  public File getProjectConfigurationDirectory() {
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultDeployProjectConfigurationConfiguration.java]: maybe just use `appEngineDirectory`?

File: src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultDeployProjectConfigurationConfiguration.java (hunk 4 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultDeployProjectConfigurationConfiguration.java]: Sometimes you gotta do what you gotta do.

File: src/main/java/com/google/cloud/tools/appengine/api/deploy/DeployProjectConfigurationConfiguration.java (hunk 3 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/api/deploy/DeployProjectConfigurationConfiguration.java]: done

File: src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultDeployProjectConfigurationConfiguration.java (hunk 5 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultDeployProjectConfigurationConfiguration.java]: yeah, that works

**Pull request merged**